### PR TITLE
Update developer guide with device_async_resource_ref guidelines

### LINKED ISSUE #15561: [DOC] Update DEVELOPER_GUIDE.md after conversion to rmm::device_async_resource_ref
## Report incorrect documentation

**Location of incorrect documentation**
DEVELOPER_GUIDE.md (libcudf)

**Describe the problems or issues found in the documentation**
After #15507 merged, the guidance on memory resource parameters in developer guide is incorrect.

**Steps taken to verify documentation is incorrect**
List any steps you have taken:

**Suggested fix for documentation**
Replace MR pointer guidance with resource_ref guidance.

### PR DESCRIPTION
**Pull request by harrism**
## Description
Closes #15561

Updates guidance in libcudf DEVELOPER_GUIDE.md to cover resource refs and change examples to not use `device_memory_resource` pointers.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -206,6 +207,27 @@ RMM provides a "default" memory resource for each device that can be accessed an
 respectively. All memory resource parameters should be defaulted to use the return value of
 `rmm::mr::get_current_device_resource()`.
 
+### Resource Refs
+
+Memory resources are passed via resource ref parameters. A resource ref is memory resource wrapper
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:212]:
```suggestion
Memory resources are passed via resource ref parameters. A resource ref is a memory resource wrapper
```

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -206,6 +207,27 @@ RMM provides a "default" memory resource for each device that can be accessed an
...
+that enables consumers to specify properties of resources that they expect. These are defined
+in the `cuda::mr` namespace of libcu++, but RMM provides some convenience wrappers in
+`rmm/resource_ref.hpp`:
+ - `rmm::device_async_resource_ref` accepts a memory resource that provides synchronous allocation
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:216]: Typo? Is this the right name?
```suggestion
 - `rmm::device_resource_ref` accepts a memory resource that provides synchronous allocation
```

**Review by pmattione-nvidia — APPROVED** [review]

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 3]
```diff
@@ -194,9 +194,10 @@ and produce `unique_ptr`s to owning objects as output. For example,
 std::unique_ptr<table> sort(table_view const& input);
 ```
 
-## rmm::device_memory_resource
+## Memory Resources (`rmm::device_memory_resource`)
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:197]:
```suggestion
## Memory Resources
```

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md]: Removed.

**Review by ttnghia — APPROVED** [review]

**Pull request merged**